File: svelte-chess/src/lib/board.ts
// Types and Constants

// Valid black pieces
export type BlackPiece = 'wR' | 'wN' | 'wQ' | 'wK' | 'wB' | 'wP';

// Valid white pieces
export type WhitePiece = 'bR' | 'bN' | 'bQ' | 'bK' | 'bB' | 'bP';

// Valid board values
export type Piece = BlackPiece | WhitePiece | null;

// Valid colours
export type Colour = 'Black' | 'White';

// Valid winning outcomes
export type WinnerColour = Colour | 'Stalemate' | null;

export type Board = Piece[][];

export type Move = {
	fromRow: number;
	fromCol: number;
	toRow: number;
	toCol: number;
};

export type Square = {
	row: number;
	col: number;
};

export const TypeMap: Record<string, string> = {
	P: 'Pawn',
	R: 'Rook',
	N: 'Knight',
	B: 'Bishop',
	Q: 'Queen',
	K: 'King'
};

// Functions

/**
 * Initialises the starting board
 * @returns A new board with the default setup
 */
export function getInitialBoard(): Board {
	return [
		['bR', 'bN', 'bB', 'bQ', 'bK', 'bB', 'bN', 'bR'],
		['bP', 'bP', 'bP', 'bP', 'bP', 'bP', 'bP', 'bP'],
		[null, null, null, null, null, null, null, null],
		[null, null, null, null, null, null, null, null],
		[null, null, null, null, null, null, null, null],
		[null, null, null, null, null, null, null, null],
		['wP', 'wP', 'wP', 'wP', 'wP', 'wP', 'wP', 'wP'],
		['wR', 'wN', 'wB', 'wQ', 'wK', 'wB', 'wN', 'wR']
	];
}

/**
 * Duplicates the contents from a board to a new object without references.
 * @param board The current board
 * @returns A new board object identical to board's contents
 */
export function cloneBoard(board: Board): Board {
	return board.map((row) => [...row]);
}

/**
 * Applies a move to the board and returns the new board state.
 * @param board The current board
 * @param move The move to apply
 * @returns A new board with the move applied
 */
export function applyMove(board: Board, move: Move): Board {
	const newBoard = cloneBoard(board);
	const piece = board[move.fromRow][move.fromCol];
	newBoard[move.toRow][move.toCol] = piece;
	newBoard[move.fromRow][move.fromCol] = null;
	return newBoard;
}

/**
 * Calculates a piece's image path.
 * @param piece The piece's board code
 * @returns The image src directory
 */
export function getPieceImagePath(piece: Piece): string {
	if (piece == null) {
		return '';
	}
	const colour = (piece[0] ?? '') === 'w' ? 'White' : 'Black';
	const type = TypeMap[piece[1]];
	return `/assets/Pieces/${colour}/${type}.png`;
}

/**
 * Checks the provided board for the lack of a king of each colour
 * @param board The current board.
 * @returns A winning state, either a colour, a draw or null
 */
export function checkWinner(board: Board): WinnerColour {
	// TODO: Add Stalemate logic
	let whiteKing = false;
	let blackKing = false;

	for (const row of board) {
		for (const piece of row) {
			if (piece === 'wK') whiteKing = true;
			if (piece === 'bK') blackKing = true;
		}
	}

	if (!whiteKing) return 'Black';
	if (!blackKing) return 'White';

	return null;
}

File: MiniMaxChessBackend/MiniMaxChessBackend/Chess/Board.cs
using Chess.Pieces;

namespace Chess {
    public class Board {
        public const int BOARDSIZE = 8;
        public Piece?[,] Tiles { get; set; }
        public List<Piece> whitePieces { get; set; }
        public List<Piece> blackPieces { get; set; }

        public Board() {
            Tiles = new Piece[BOARDSIZE, BOARDSIZE];
            whitePieces = new List<Piece>();
            blackPieces = new List<Piece>();
        }

        public void Initialise() {
            // Clear board
            for (int i = 0; i < BOARDSIZE; i++) {
                for (int j = 0; j < BOARDSIZE; j++) {
                    Tiles[i, j] = null;
                }
            }

            // Computer Black Pieces
            Tiles[0, 0] = new Rook(PieceColour.Black);
            Tiles[0, 1] = new Knight(PieceColour.Black);
            Tiles[0, 2] = new Bishop(PieceColour.Black);
            Tiles[0, 3] = new Queen(PieceColour.Black);
            Tiles[0, 4] = new King(PieceColour.Black);
            Tiles[0, 5] = new Bishop(PieceColour.Black);
            Tiles[0, 6] = new Knight(PieceColour.Black);
            Tiles[0, 7] = new Rook(PieceColour.Black);

            for (int j = 0; j < BOARDSIZE; j++) {
                Tiles[1, j] = new Pawn(PieceColour.Black);
            }

            // Player White pieces
            Tiles[7, 0] = new Rook(PieceColour.White);
            Tiles[7, 1] = new Knight(PieceColour.White);
            Tiles[7, 2] = new Bishop(PieceColour.White);
            Tiles[7, 3] = new Queen(PieceColour.White);
            Tiles[7, 4] = new King(PieceColour.White);
            Tiles[7, 5] = new Bishop(PieceColour.White);
            Tiles[7, 6] = new Knight(PieceColour.White);
            Tiles[7, 7] = new Rook(PieceColour.White);

            for (int j = 0; j < BOARDSIZE; j++) {
                Tiles[6, j] = new Pawn(PieceColour.White);
            }
        }

        public bool IsInsideBoard(int row, int col) {
            return row >= 0 && row < BOARDSIZE && col >= 0 && col < BOARDSIZE;
        }

        public void MakeMove(Move move) {
            Piece? piece = Tiles[move.FromRow, move.FromCol];

            if (piece == null) {
                return;
            }

            Piece? takenPiece = Tiles[move.ToRow, move.ToCol];

            if (takenPiece != null) {
                if (takenPiece.Colour == PieceColour.White) {
                    whitePieces.Add(takenPiece);
                } else {
                    blackPieces.Add(takenPiece);
                }
            }

            Tiles[move.ToRow, move.ToCol] = piece;
            Tiles[move.FromRow, move.FromCol] = null;
        }

        public Board Clone() {
            Board newBoard = new Board();
            for (int i = 0; i < BOARDSIZE; i++) {
                for (int j = 0; j < BOARDSIZE; j++) {
                    if (Tiles[i, j] != null) {
                        newBoard.Tiles[i, j] = Tiles[i, j].Clone();
                    } else {
                        newBoard.Tiles[i, j] = null;
                    }
                }
            }

            return newBoard;
        }

        public List<Move> GenerateAllMoves(PieceColour colour) {
            List<Move> moves = new List<Move>();

            for (int i = 0; i < BOARDSIZE; i++) {
                for (int j = 0; j < BOARDSIZE; j++) {
                    Piece? piece = Tiles[i, j];
                    if (piece != null && piece.Colour == colour) {
                        moves.AddRange(piece.GetLegalMoves(this, i, j));
                    }
                }
            }

            return moves;
        }

        public bool IsKingAlive(PieceColour colour) {
            for (int i = 0; i < BOARDSIZE; i++) {
                for (int j = 0; j < BOARDSIZE; j++) {
                    Piece? piece = Tiles[i, j];
                    if (piece != null && piece.Type == PieceType.King && piece.Colour == colour) {
                        return true;
                    }
                }
            }

            return false;
        }

        public int Evaluate(PieceColour colour) {
            int score = 0;

            for (int i = 0; i < BOARDSIZE; i++) {
                for (int j = 0; j < BOARDSIZE; j++) {
                    Piece? piece = Tiles[i, j];

                    if (piece != null) {
                        int pieceValue = GetPieceValue(piece.Type);

                        if (piece.Colour == colour) {
                            score += pieceValue;
                        } else {
                            score -= pieceValue;
                        }
                    }
                }
            }

            return score;
        }

        public int GetPieceValue(PieceType type) {
            switch (type) {
                case PieceType.Pawn: {
                    return 10;
                }

                case PieceType.Knight: {
                    return 30;
                }

                case PieceType.Bishop: {
                    return 30;
                }

                case PieceType.Rook: {
                    return 50;
                }

                case PieceType.Queen: {
                    return 90;
                }

                case PieceType.King: {
                    return 900;
                }

                default: {
                    return 0;
                }
            }
        }
    }
}
